envelope.go

```diff
@@ -67,14 +67,24 @@ func (w *envelopeWriter) Marshal(message any) *Error {
 		}
 		return nil
 	}
-	raw, err := w.codec.Marshal(message)
+
+	buffer := w.bufferPool.Get()
+	defer w.bufferPool.Put(buffer)
+
+	// Reuse byte buffer if codec supports MarshalAppend.
+	var (
+		raw []byte
+		err error
+	)
+	if c, ok := w.codec.(marshalAppend); ok {
+		raw, err = c.MarshalAppend(buffer.Bytes(), message)
+	} else {
+		raw, err = w.codec.Marshal(message)
```

Comment:
I think performance along this path has gotten worse, since we no longer put `raw` into the pool when we're done with it. Adding a bit of code to the `else` block should fix it:

```go
if len(raw) > 0 {
  defer w.bufferPool.Put(bytes.NewBuffer(raw))
}
```

Reply:
Even with this change, I think we're failing to effectively pool here. At base, the buffers in the pool are pretty small. `MarshalAppend` is likely to grow the slice, in which case returning `buffer` to the pool no longer helps much. Give me a moment to play with this a bit locally and find a way to fix this that isn't _too_ prone to bugs down the road.

Reply:
I think it will do the right thing - doing `buffer.Write(raw)` will grow the buffer as needed, so it will be returned to the pool with a larger size. I think we can remove adding an additional bytes.Buffer to the pool.